lib/matplotlib/tests/test_axes.py

```diff
@@ -4603,35 +4603,57 @@ def test_hist_stacked_bar():
     ax.legend(loc='upper right', bbox_to_anchor=(1.0, 1.0), ncols=1)
 
 
-@pytest.mark.parametrize("histtype", ["step", "stepfilled"])
-@pytest.mark.parametrize("color", [["blue", "green", "brown"], None])
-@pytest.mark.parametrize("edgecolor", [["red", "black", "blue"], [None]*3])
-@pytest.mark.parametrize("facecolor", [["blue", "green", "brown"], [None]*3])
+@pytest.mark.parametrize("histtype, color, edgecolor, facecolor,",
+                         [("step", ["b", "g", "m"], ["r", "k", "b"], ["b", "g", "m"]),
+                          ("stepfilled", ["b", "g", "m"], ["r", "k", "b"], [None]*3),
```

Comment:
> > None as Ill have to iterate through it?
> > Enumerate should I think repeat scaler values
> 
> And https://docs.pytest.org/en/7.1.x/how-to/assert.html

Seems like enumerate does not repeat scalar values

Reply:
Which part of code exactly are we testing with the color variations? Fundamentally, we should not have to run full plots to ensure color precedence, that should not be the responsibility of `hist()` itself. Maybe we need something like in the case of scatter? 

https://github.com/matplotlib/matplotlib/blob/b3d29fb036b0d9de2fb66ee7bcad6887654b3706/lib/matplotlib/axes/_axes.py#L4543

Extracting such a function allows to test a complex interplay of parameters indepentent of the plotting function. Testing such a function is fast and can have the whole product space of parameters. Then, one or two variants for drawing a hist should be enough.

Reply:
```
    if(histtype == "step"):
        assert [ax.patches[i].get_edgecolor() for i in range(len(patches))] == color if color else edgecolor
```
So the plan is to just check a bunch of assertions like this for facecolor edgecolor? and plot only for the other parameters like hatch, linewidth etc.?

Reply:
> Where the signature on a color setting function would be something like:
> 
> ```python
> def _hist_color_params(histtype,  color, facecolor, edgecolor, hatchcolor?):
> #Color setting logic
> 
> return facecolor, edgecolor, hatch color?
> ```
> 
> I don't know if we allow setting hatchcolor, but it's a private function so param space can grow, also technically can maybe do just the params that are also args into hist and everything else is a kwarg so that unpacking happens inside.
> 
> Also, list comprehension ,so something like:
> 
> `assert all(p.get_edgecolor() == color for p in histplot.patches)`
> 
> not ax.patchee cause that I think includes the axes background color

So the plan is to make this function, use it in def hist(...) and more or less put all of these lines:
```
        if histtype == "step":
            edgecolors = itertools.cycle(np.atleast_1d(kwargs.get('edgecolor',
                                                                  colors)))
        else:
            edgecolors = itertools.cycle(np.atleast_1d(kwargs.get("edgecolor", None)))

        facecolors = itertools.cycle(np.atleast_1d(kwargs.get('facecolor', colors)))
        hatches = itertools.cycle(np.atleast_1d(kwargs.get('hatch', None)))
        linewidths = itertools.cycle(np.atleast_1d(kwargs.get('linewidth', None)))
        linestyles = itertools.cycle(np.atleast_1d(kwargs.get('linestyle', None)))
```
inside there? Call the function at the beginning of the def hist(...) function and use those values? Definitely feel like opening a new PR for this. Although Im thinking if the function is even necessary.....

Reply:
The tests will include the ones in the table

Reply:
Do I include this in this PR or another one?